ray-curator/ray_curator/utils/file_utils.py

```diff
@@ -0,0 +1,372 @@
+import os
+import warnings
+from multiprocessing import Pool
+from typing import Literal
+
+import fsspec
+import pyarrow.parquet as pq
+from fsspec.core import get_filesystem_class, split_protocol
+from fsspec.parquet import open_parquet_file
+from loguru import logger
+
+
+def get_fs(path: str, storage_options: dict[str, str] | None = None) -> fsspec.AbstractFileSystem:
+    if not storage_options:
+        storage_options = {}
+    protocol, path = split_protocol(path)
+    return get_filesystem_class(protocol)(**storage_options)
+
+
+def filter_files_by_extension(
+    files_list: list[str],
+    keep_extensions: str | list[str],
+) -> list[str]:
+    filtered_files = []
+    if isinstance(keep_extensions, str):
+        keep_extensions = [keep_extensions]
+
+    # Ensure that the extensions are prefixed with a dot
+    file_extensions = tuple([s if s.startswith(".") else f".{s}" for s in keep_extensions])
+
+    for file in files_list:
+        if file.endswith(file_extensions):
+            filtered_files.append(file)
+
+    if len(files_list) != len(filtered_files):
+        warnings.warn("Skipped at least one file due to unmatched file extension(s).", stacklevel=2)
+
+    return filtered_files
+
+
+def get_all_files_paths_under(
+    input_dir: str,
+    recurse_subdirectories: bool = False,
+    keep_extensions: str | list[str] | None = None,
+    storage_options: dict | None = None,
+    fs: fsspec.AbstractFileSystem | None = None,
+) -> list[str]:
+    # TODO: update with a more robust fsspec method
+    if fs is None:
+        fs = get_fs(input_dir, storage_options)
+
+    file_ls = fs.find(input_dir, maxdepth=None if recurse_subdirectories else 1)
+    if "://" in input_dir:
+        protocol = input_dir.split("://")[0]
+        file_ls = [f"{protocol}://{f}" for f in file_ls]
+
+    file_ls.sort()
+    if keep_extensions is not None:
+        file_ls = filter_files_by_extension(file_ls, keep_extensions)
+    return file_ls
+
+
+def remove_and_create_dir(
```

Comment:
I think this function (and all others below it), are not being used anywhere. Should we still keep them in this PR?

I remember several being used for dedupe, maybe we should only re-add them as we add modules that use them?

Reply:
Happy to remove them for now and re-add them as they are needed.